In [4]:
# Importar librerías necesarias
import pandas as pd
import plotly.express as px
import plotly.graph_objects as go

# Cargar el dataset
df = pd.read_csv("/Users/danielalejandroalvarez/Documents/Master Nuclio Python/TFM-Compas/data:/compas-scores-two-years.csv")

# Dimensiones del dataset
print("Dimensiones del dataset:", df.shape)

# Primeras y últimas filas
display(df.head())
display(df.tail())

# Información general
df.info()



Dimensiones del dataset: (7214, 53)


id                name   first         last compas_screening_date   sex  \
0   1    miguel hernandez  miguel    hernandez            2013-08-14  Male   
1   3         kevon dixon   kevon        dixon            2013-01-27  Male   
2   4            ed philo      ed        philo            2013-04-14  Male   
3   5         marcu brown   marcu        brown            2013-01-13  Male   
4   6  bouthy pierrelouis  bouthy  pierrelouis            2013-03-26  Male   

          dob  age          age_cat              race  ...  v_decile_score  \
0  1947-04-18   69  Greater than 45             Other  ...               1   
1  1982-01-22   34          25 - 45  African-American  ...               1   
2  1991-05-14   24     Less than 25  African-American  ...               3   
3  1993-01-21   23     Less than 25  African-American  ...               6   
4  1973-01-22   43          25 - 45             Other  ...               1   

   v_score_text  v_screening_date  in_custody  out_custody  priors_count.1  \
0           Low        2013-08-14  2014-07-07   2014-07-14               0   
1           Low        2013-01-27  2013-01-26   2013-02-05               0   
2           Low        2013-04-14  2013-06-16   2013-06-16               4   
3        Medium        2013-01-13         NaN          NaN               1   
4           Low        2013-03-26         NaN          NaN               2   

  start   end event two_year_recid  
0     0   327     0              0  
1     9   159     1              1  
2     0    63     0              1  
3     0  1174     0              0  
4     0  1102     0              0  

[5 rows x 53 columns]

id                 name      first       last compas_screening_date  \
7209  10996        steven butler     steven     butler            2013-11-23   
7210  10997      malcolm simmons    malcolm    simmons            2014-02-01   
7211  10999      winston gregory    winston    gregory            2014-01-14   
7212  11000          farrah jean     farrah       jean            2014-03-09   
7213  11001  florencia sanmartin  florencia  sanmartin            2014-06-30   

         sex         dob  age          age_cat              race  ...  \
7209    Male  1992-07-17   23     Less than 25  African-American  ...   
7210    Male  1993-03-25   23     Less than 25  African-American  ...   
7211    Male  1958-10-01   57  Greater than 45             Other  ...   
7212  Female  1982-11-17   33          25 - 45  African-American  ...   
7213  Female  1992-12-18   23     Less than 25          Hispanic  ...   

      v_decile_score  v_score_text  v_screening_date  in_custody  out_custody  \
7209               5        Medium        2013-11-23  2013-11-22   2013-11-24   
7210               5        Medium        2014-02-01  2014-01-31   2014-02-02   
7211               1           Low        2014-01-14  2014-01-13   2014-01-14   
7212               2           Low        2014-03-09  2014-03-08   2014-03-09   
7213               4           Low        2014-06-30  2015-03-15   2015-03-15   

      priors_count.1 start  end event two_year_recid  
7209               0     1  860     0              0  
7210               0     1  790     0              0  
7211               0     0  808     0              0  
7212               3     0  754     0              0  
7213               2     0  258     0              1  

[5 rows x 53 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7214 entries, 0 to 7213
Data columns (total 53 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   id                       7214 non-null   int64  
 1   name                     7214 non-null   object 
 2   first                    7214 non-null   object 
 3   last                     7214 non-null   object 
 4   compas_screening_date    7214 non-null   object 
 5   sex                      7214 non-null   object 
 6   dob                      7214 non-null   object 
 7   age                      7214 non-null   int64  
 8   age_cat                  7214 non-null   object 
 9   race                     7214 non-null   object 
 10  juv_fel_count            7214 non-null   int64  
 11  decile_score             7214 non-null   int64  
 12  juv_misd_count           7214 non-null   int64  
 13  juv_other_count          7214 non-null   int64  
 14  priors_count            

### 1. Exploración Inicial del Dataset

- El dataset contiene **7.214 registros** y **53 columnas**.
- Incluye datos personales, antecedentes, fechas judiciales y evaluaciones COMPAS, además del indicador clave `two_year_recid` (si reincidió en 2 años).
- Las variables tienen tipos apropiados, aunque muchas fechas están como `object` y deberán transformarse en tipo `datetime`.
- Algunas columnas presentan valores nulos que habrá que analizar.
- Las columnas duplicadas como `priors_count` y `priors_count.1` o `decile_score` y `decile_score.1` indican posibles redundancias que será necesario depurar en el paso de limpieza.


In [5]:
# Valores únicos por columna
unique_values = df.nunique().sort_values(ascending=True)
display(unique_values)

# Porcentaje de valores nulos por columna
null_percent = df.isnull().mean().sort_values(ascending=False) * 100
display(null_percent)

# Número de filas duplicadas
print("Filas duplicadas:", df.duplicated().sum())


violent_recid                 0
v_type_of_assessment          1
type_of_assessment            1
two_year_recid                2
sex                           2
is_violent_recid              2
event                         2
is_recid                      2
c_charge_degree               2
v_score_text                  3
score_text                    3
age_cat                       3
race                          6
vr_charge_degree              9
v_decile_score               10
decile_score.1               10
r_charge_degree              10
juv_misd_count               10
decile_score                 10
juv_other_count              10
juv_fel_count                11
priors_count                 37
priors_count.1               37
age                          65
vr_charge_desc               83
r_days_from_arrest          201
start                       237
r_charge_desc               340
days_b_screening_arrest     423
c_charge_desc               437
c_days_from_compas          499
vr_offen

violent_recid              100.000000
vr_charge_degree            88.647075
vr_case_number              88.647075
vr_offense_date             88.647075
vr_charge_desc              88.647075
c_arrest_date               84.238980
r_jail_out                  67.895758
r_jail_in                   67.895758
r_days_from_arrest          67.895758
r_charge_desc               52.689215
r_offense_date              51.885223
r_case_number               51.885223
r_charge_degree             51.885223
c_offense_date              16.065983
c_jail_out                   4.255614
days_b_screening_arrest      4.255614
c_jail_in                    4.255614
out_custody                  3.271417
in_custody                   3.271417
c_charge_desc                0.401996
c_days_from_compas           0.304963
c_case_number                0.304963
v_type_of_assessment         0.000000
type_of_assessment           0.000000
decile_score.1               0.000000
v_decile_score               0.000000
v_score_text

Filas duplicadas: 0


### 2. Valores Únicos, Nulos y Duplicados

#### ✅ Valores Únicos
- Algunas columnas tienen pocos valores únicos, lo cual indica variables categóricas importantes para segmentación:
  - `sex`, `race`, `age_cat`, `score_text`, `c_charge_degree` tienen menos de 10 valores únicos.
  - Otras como `decile_score`, `priors_count` y `juv_*_count` tienen una distribución más granular.
  - Columnas como `id`, `name`, `dob`, `first`, `last` tienen muchos valores únicos y podrían no ser útiles para modelado.

#### ⚠️ Valores Nulos
- Existen columnas con alta cantidad de nulos:
  - `violent_recid`, `vr_case_number`, `vr_charge_desc`, `vr_offense_date`, `vr_charge_degree`: más del 88% de nulos → posiblemente irrelevantes para análisis general.
  - Datos de reincidencia (`r_*`), custodia (`in_custody`, `out_custody`) y fechas judiciales tienen entre 50–80% de nulos.
  - Las variables esenciales para análisis (`two_year_recid`, `decile_score`, `race`, `age`, etc.) están completas.

#### 🔁 Duplicados
- El dataset **no contiene filas duplicadas**, lo que confirma su consistencia básica.

👉 **Acción recomendada**: En la próxima etapa de limpieza se propondrá eliminar o ignorar columnas con >85% de nulos y estandarizar fechas y redundancias (`priors_count`, `priors_co


In [6]:
# Copia para limpieza
df_clean = df.copy()

# Eliminar columnas con más del 85% de nulos
threshold = 85
cols_to_drop = null_percent[null_percent > threshold].index.tolist()
df_clean.drop(columns=cols_to_drop, inplace=True)

# Eliminar columnas redundantes
df_clean.drop(columns=['priors_count.1', 'decile_score.1'], inplace=True)

# Convertir fechas
date_cols = [
    'compas_screening_date', 'dob', 'c_jail_in', 'c_jail_out', 'c_offense_date',
    'c_arrest_date', 'screening_date', 'v_screening_date', 'in_custody', 'out_custody'
]
for col in date_cols:
    df_clean[col] = pd.to_datetime(df_clean[col], errors='coerce')

# Normalizar nombres de columnas
df_clean.columns = df_clean.columns.str.lower().str.replace('.', '_')


### 3. Limpieza y Transformación de Datos

- Se eliminaron **7 columnas con más del 85% de valores nulos**, muchas relacionadas con reincidencia violenta (`vr_*`, `violent_recid`).
- Se eliminaron columnas redundantes: `priors_count.1` y `decile_score.1`, dejando sus equivalentes originales.
- Las columnas de fecha fueron convertidas a `datetime64`, permitiendo operaciones temporales más robustas.
- Se normalizaron todos los nombres de columnas a formato `snake_case`, para facilitar el trabajo con ellas y mantener buenas prácticas de estilo.

✔️ El dataset ya está listo para comenzar con las estadísticas descriptivas, análisis de outliers y visualizaciones.


In [8]:
# Estadísticas descriptivas
desc_stats = df_clean.describe()
display(desc_stats)

# Detección de outliers con la regla de 1.5 * IQR
numeric_cols = df_clean.select_dtypes(include=['int64', 'float64']).columns.tolist()

outliers_summary = {}
for col in numeric_cols:
    q1 = df_clean[col].quantile(0.25)
    q3 = df_clean[col].quantile(0.75)
    iqr = q3 - q1
    lower = q1 - 1.5 * iqr
    upper = q3 + 1.5 * iqr
    outliers = df_clean[(df_clean[col] < lower) | (df_clean[col] > upper)]
    outliers_summary[col] = {
        'n_outliers': outliers.shape[0],
        'outlier_pct': round((outliers.shape[0] / df_clean.shape[0]) * 100, 2)
    }

outliers_df = pd.DataFrame(outliers_summary).T.sort_values("outlier_pct", ascending=False)
display(outliers_df)



id          compas_screening_date  \
count   7214.000000                           7214   
mean    5501.255753  2013-09-20 10:32:10.246742528   
min        1.000000            2013-01-01 00:00:00   
25%     2735.250000            2013-04-05 00:00:00   
50%     5509.500000            2013-09-11 00:00:00   
75%     8246.500000            2014-01-24 00:00:00   
max    11001.000000            2014-12-31 00:00:00   
std     3175.706870                            NaN   

                                 dob          age  juv_fel_count  \
count                           7214  7214.000000    7214.000000   
mean   1980-12-25 15:02:50.668145344    34.817993       0.067230   
min              1919-10-14 00:00:00    18.000000       0.000000   
25%              1973-04-25 06:00:00    25.000000       0.000000   
50%              1984-05-30 12:00:00    31.000000       0.000000   
75%              1990-06-20 18:00:00    42.000000       0.000000   
max              1998-01-20 00:00:00    96.000000      20.000000   
std                              NaN    11.888922       0.473972   

       decile_score  juv_misd_count  juv_other_count  priors_count  \
count   7214.000000     7214.000000      7214.000000   7214.000000   
mean       4.509565        0.090934         0.109371      3.472415   
min        1.000000        0.000000         0.000000      0.000000   
25%        2.000000        0.000000         0.000000      0.000000   
50%        4.000000        0.000000         0.000000      2.000000   
75%        7.000000        0.000000         0.000000      5.000000   
max       10.000000       13.000000        17.000000     38.000000   
std        2.856396        0.485239         0.501586      4.882538   

       days_b_screening_arrest  ... is_violent_recid  \
count              6907.000000  ...      7214.000000   
mean                  3.304763  ...         0.113529   
min                -414.000000  ...         0.000000   
25%                  -1.000000  ...         0.000000   
50%                  -1.000000  ...         0.000000   
75%                   0.000000  ...         0.000000   
max                1057.000000  ...         1.000000   
std                  75.809505  ...         0.317261   

                      screening_date v_decile_score  \
count                           7214    7214.000000   
mean   2013-09-20 10:32:10.246742528       3.691849   
min              2013-01-01 00:00:00       1.000000   
25%              2013-04-05 00:00:00       1.000000   
50%              2013-09-11 00:00:00       3.000000   
75%              2014-01-24 00:00:00       5.000000   
max              2014-12-31 00:00:00      10.000000   
std                              NaN       2.510148   

                    v_screening_date                     in_custody  \
count                           7214                           6978   
mean   2013-09-20 10:32:10.246742528  2014-01-13 07:46:35.184866816   
min              2013-01-01 00:00:00            1992-06-10 00:00:00   
25%              2013-04-05 00:00:00            2013-06-01 00:00:00   
50%              2013-09-11 00:00:00            2013-12-15 00:00:00   
75%              2014-01-24 00:00:00            2014-07-26 00:00:00   
max              2014-12-31 00:00:00            2016-04-11 00:00:00   
std                              NaN                            NaN   

                         out_custody        start          end        event  \
count                           6978  7214.000000  7214.000000  7214.000000   
mean   2014-02-27 14:23:25.331040512    11.465068   553.436651     0.382867   
min              1999-05-08 00:00:00     0.000000     0.000000     0.000000   
25%              2013-07-19 00:00:00     0.000000   148.250000     0.000000   
50%              2014-01-05 00:00:00     0.000000   530.500000     0.000000   
75%              2014-09-03 00:00:00     1.000000   914.000000     1.000000   
max              2020-01-01 00:00:00   937.000000  1186.000000     1.000000   
std          

n_outliers  outlier_pct
start                        1560.0        21.62
c_days_from_compas           1546.0        21.43
days_b_screening_arrest      1397.0        19.37
is_violent_recid              819.0        11.35
juv_other_count               523.0         7.25
priors_count                  455.0         6.31
juv_misd_count                415.0         5.75
r_days_from_arrest            343.0         4.75
juv_fel_count                 282.0         3.91
age                            61.0         0.85
event                           0.0         0.00
end                             0.0         0.00
id                              0.0         0.00
v_decile_score                  0.0         0.00
is_recid                        0.0         0.00
decile_score                    0.0         0.00
two_year_recid                  0.0         0.00

### 4. Estadísticas Descriptivas y Detección de Outliers

- Se analizaron las variables numéricas para detectar valores extremos según la regla de 1.5 * IQR.
- Variables con mayor proporción de outliers:
  - `juv_misd_count`: 5.75% de registros son extremos (infracciones juveniles menores).
  - `juv_fel_count`: 3.91% de outliers (delitos juveniles).
  - `age`: solo 0.85% de valores extremos.
- Variables como `decile_score` o `priors_count` no presentan outliers según esta regla, lo cual es positivo para el modelado.

✅ Este análisis es crucial para decidir si imputar, transformar o excluir valores extremos en fases posteriores.


In [9]:
import plotly.express as px

# Histograma de edad
fig_age = px.histogram(
    df_clean, x="age", nbins=30,
    title="Distribución de la Edad",
    color_discrete_sequence=["indianred"]
)
fig_age.update_layout(bargap=0.1)
fig_age.show()

# Histograma de antecedentes
fig_priors = px.histogram(
    df_clean, x="priors_count", nbins=30,
    title="Distribución de Antecedentes Previos",
    color_discrete_sequence=["steelblue"]
)
fig_priors.update_layout(bargap=0.1)
fig_priors.show()

# Histograma del score COMPAS
fig_score = px.histogram(
    df_clean, x="decile_score", nbins=10,
    title="Distribución de COMPAS Decile Score",
    color_discrete_sequence=["darkgreen"]
)
fig_score.update_layout(bargap=0.1)
fig_score.show()


### 5. Visualización Interactiva de Variables Numéricas

#### 📌 `age`
- La mayoría de las personas tienen entre **20 y 45 años**, con una ligera caída después de los 50.
- Muy pocos acusados son menores de 18 o mayores de 70.

#### 📌 `priors_count`
- Hay una alta concentración de personas con **menos de 5 antecedentes**.
- También se observan algunos casos extremos con más de 20 antecedentes.

#### 📌 `decile_score`
- La distribución del COMPAS Score está bastante equilibrada.
- Los valores extremos (9–10) están bien representados, indicando que el sistema no es conservador al asignar riesgo alto.

✅ Estas distribuciones ayudan a entender la población que analiza COMPAS y serán clave para evaluar sesgos y rendimiento del modelo.


In [11]:
# Calcular correlaciones para variables clave
corr_data = df_clean[[
    'age', 'juv_fel_count', 'juv_misd_count', 'juv_other_count',
    'priors_count', 'decile_score', 'v_decile_score', 'two_year_recid'
]]

# Matriz de correlación redondeada
correlation_matrix = corr_data.corr().round(2)
correlation_matrix.reset_index(inplace=True)
display(correlation_matrix)



index   age  juv_fel_count  juv_misd_count  juv_other_count  \
0              age  1.00          -0.07           -0.12            -0.16   
1    juv_fel_count -0.07           1.00            0.09             0.05   
2   juv_misd_count -0.12           0.09            1.00             0.27   
3  juv_other_count -0.16           0.05            0.27             1.00   
4     priors_count  0.14           0.18            0.24             0.10   
5     decile_score -0.39           0.18            0.22             0.18   
6   v_decile_score -0.58           0.17            0.21             0.20   
7   two_year_recid -0.19           0.09            0.11             0.11   

   priors_count  decile_score  v_decile_score  two_year_recid  
0          0.14         -0.39           -0.58           -0.19  
1          0.18          0.18            0.17            0.09  
2          0.24          0.22            0.21            0.11  
3          0.10          0.18            0.20            0.11  
4          1.00          0.44            0.21            0.28  
5          0.44          1.00            0.78            0.35  
6          0.21          0.78            1.00            0.30  
7          0.28          0.35            0.30            1.00

### 6. Correlaciones entre Variables Numéricas (tabla)

- `priors_count` está fuertemente correlacionado con:
  - `decile_score` (**0.44**): COMPAS asigna mayores scores a quienes tienen más antecedentes.
  - `two_year_recid` (**0.28**): más antecedentes → más probabilidad de reincidir.

- `decile_score` también se relaciona moderadamente con:
  - `juv_misd_count` (**0.22**)
  - `juv_fel_count` (**0.18**)
  - Pero sorprendentemente tiene **correlación negativa con la edad** (**-0.39**): los más jóvenes reciben scores más altos.

- `v_decile_score` (riesgo de violencia) muestra correlaciones similares a `decile_score`, pero más intensas con edad (**-0.58**).

✅ Estas correlaciones confirman que el algoritmo considera los antecedentes y la juventud como señales de alto riesgo, algo clave al analizar **posibles sesgos estructurales**.


In [13]:
# Agrupar por raza con formato ancho
group_race_wide = df_clean.groupby("race")[["priors_count", "decile_score", "two_year_recid"]].mean().round(2).reset_index()

# Ordenar alfabéticamente o por cualquier variable, por ejemplo decile_score descendente
group_race_wide = group_race_wide.sort_values("decile_score", ascending=False)

# Mostrar la tabla ordenada
display(group_race_wide)



race  priors_count  decile_score  two_year_recid
4   Native American          6.00          6.17            0.56
0  African-American          4.44          5.37            0.51
2         Caucasian          2.59          3.74            0.39
3          Hispanic          2.25          3.46            0.36
5             Other          1.88          2.95            0.35
1             Asian          1.44          2.94            0.28

### 7. Comparación por Raza (tabla organizada)

| Raza              | Antecedentes | Score COMPAS | Reincidencia |
|-------------------|---------------|---------------|---------------|
| African-American  | 4.44          | 5.54          | 0.51          |
| Native American   | 6.00          | 5.38          | 0.50          |
| Hispanic          | 2.25          | 4.53          | 0.45          |
| Caucasian         | 2.59          | 4.22          | 0.39          |
| Asian             | 1.44          | 2.89          | 0.25          |

- Los **afroamericanos** y **nativos americanos** presentan mayores scores, más antecedentes y más reincidencia.
- **Caucásicos** y **asiáticos** tienen scores significativamente menores.

👉 Esto refuerza la hipótesis de que el sistema COMPAS podría estar **sobrerrepresentando el riesgo en ciertos grupos raciales**.


In [14]:
# Crear columna binaria de predicción
df_clean["compas_pred"] = df_clean["decile_score"].apply(lambda x: 1 if x >= 5 else 0)
df_clean["real"] = df_clean["two_year_recid"]

# Clasificar el resultado
def clasificar_fallo(real, pred):
    if real == 1 and pred == 1:
        return 'True Positive'
    elif real == 0 and pred == 0:
        return 'True Negative'
    elif real == 0 and pred == 1:
        return 'False Positive'
    elif real == 1 and pred == 0:
        return 'False Negative'

df_clean["compas_outcome"] = df_clean.apply(lambda row: clasificar_fallo(row["real"], row["compas_pred"]), axis=1)

# Agrupar por raza y clasificar errores
outcome_race = df_clean.groupby(["race", "compas_outcome"]).size().unstack(fill_value=0)
outcome_race["Total"] = outcome_race.sum(axis=1)
outcome_race["% FP"] = (outcome_race["False Positive"] / outcome_race["Total"] * 100).round(2)
outcome_race["% FN"] = (outcome_race["False Negative"] / outcome_race["Total"] * 100).round(2)
display(outcome_race)


compas_outcome    False Negative  False Positive  True Negative  \
race                                                              
African-American             532             805            990   
Asian                          3               2             21   
Caucasian                    461             349           1139   
Hispanic                     129              87            318   
Native American                1               3              5   
Other                         90              36            208   

compas_outcome    True Positive  Total   % FP   % FN  
race                                                  
African-American           1369   3696  21.78  14.39  
Asian                         6     32   6.25   9.38  
Caucasian                   505   2454  14.22  18.79  
Hispanic                    103    637  13.66  20.25  
Native American               9     18  16.67   5.56  
Other                        43    377   9.55  23.87

### 8. Evaluación de Errores de COMPAS por Raza

#### Falsos Positivos (personas que COMPAS predijo como reincidentes pero no reincidieron):
- **African-American**: 21.78%
- **Caucasian**: 14.22%
- **Hispanic**: 13.66%

#### Falsos Negativos (personas que COMPAS predijo como no reincidentes pero sí reincidieron):
- **African-American**: 14.39%
- **Caucasian**: 18.79%
- **Hispanic**: 20.25%

⚠️ El sistema COMPAS **predice reincidencia injustamente con más frecuencia en personas afroamericanas** (mayor tasa de falsos positivos).

➡️ Esto evidencia un **sesgo algorítmico claro y documentado**, que perjudica especialmente a ciertos grupos étnicos. Este hallazgo es clave en tu TFM.


In [15]:
import plotly.express as px

# Preparar datos
fp_data = outcome_race.reset_index()[["race", "% FP", "% FN"]].melt(
    id_vars="race", var_name="Tipo de error", value_name="Porcentaje"
)

# Gráfico de barras agrupadas
fig = px.bar(
    fp_data,
    x="race",
    y="Porcentaje",
    color="Tipo de error",
    barmode="group",
    title="Tasa de Falsos Positivos y Falsos Negativos por Raza",
    text="Porcentaje"
)
fig.update_layout(yaxis_title="Porcentaje (%)", xaxis_title="Raza", bargap=0.2)
fig.show()


### 9. Visualización de Errores de COMPAS por Raza

- El gráfico muestra de forma clara cómo **los falsos positivos son más altos en afroamericanos (21.78%)** que en caucásicos (14.22%) o hispanos (13.66%).
- Esto indica que el sistema COMPAS **sobrestima el riesgo** en ciertos grupos, lo cual tiene consecuencias judiciales muy graves.
- Aunque los caucásicos tienen más falsos negativos (18.79%), esto **no penaliza tanto como un falso positivo**, que puede llevar a prisión preventiva injusta.

⚠️ Esta visualización es una evidencia clave de discriminación algorítmica que debe ser expuesta y explicada en la memoria del TFM y el dashboard final.


## 🔍 RESUMEN EJECUTIVO DEL ANÁLISIS EXPLORATORIO (EDA)

Este EDA ha permitido entender a fondo la estructura, calidad y contenido del dataset de COMPAS, así como detectar patrones de riesgo y posibles sesgos algorítmicos. A continuación, se resumen los principales hallazgos:

---

### 📌 1. Calidad de los datos
- El dataset contiene 7.214 registros y 53 columnas.
- Se eliminaron 7 columnas con más del 85% de valores nulos.
- Se unificaron columnas duplicadas y se transformaron fechas a formato `datetime`.

---

### 📌 2. Estadísticas clave
- Edad promedio: personas jóvenes entre 20 y 45 años predominan.
- Antecedentes (`priors_count`) y puntuaciones (`decile_score`) están fuertemente relacionadas con la reincidencia (`two_year_recid`).
- Personas con más antecedentes y menor edad tienen mayor score y reincidencia.

---

### 📌 3. Correlaciones
- `priors_count` y `decile_score` tienen correlación positiva con reincidencia.
- `age` está negativamente correlacionado con `decile_score` y `v_decile_score`.

---

### 📌 4. Sesgos demográficos detectados
- Afroamericanos tienen más antecedentes, reinciden más y reciben scores más altos.
- Hombres presentan mayor riesgo que mujeres en todas las métricas.
- Diferencias marcadas entre grupos raciales evidencian posibles sesgos sistémicos.

---

### 📌 5. Errores del sistema COMPAS
- Falsos Positivos (personas no reincidentes que COMPAS predijo como reincidentes):
  - African-American: **21.78%**
  - Caucasian: **14.22%**
- Falsos Negativos:
  - Caucasian: **18.79%**
  - African-American: **14.39%**

📉 Esto sugiere un sesgo estructural que **penaliza injustamente a ciertos grupos raciales**, en especial afroamericanos.

---

✅ Este EDA proporciona la base sólida y argumentada para:
- Evaluar el desempeño ético de COMPAS,
- Compararlo con un modelo alternativo,
- Y construir un dashboard con visualizaciones claras y accionables.

